File: test/support/normalization.py
from khiva.array import *
from khiva.library import *
from khiva.normalization import *

set_backend(KHIVABackend.KHIVA_BACKEND_CPU)

a = Array.from_list([1, 2, 3, 4, 5], dtype.f32)

print('decimal_scaling_norm')
print(decimal_scaling_norm(a).to_list())
print()

print('decimal_scaling_norm_in_place')
b = a.copy()
decimal_scaling_norm_in_place(b)
print(b.to_list())
print()

print('max_min_norm')
print(max_min_norm(a).to_list())
print()

print('max_min_norm_in_place')
b = a.copy()
max_min_norm_in_place(b)
print(b.to_list())
print()

print('mean_norm')
print(mean_norm(a).to_list())
print()

print('mean_norm_in_place')
b = a.copy()
mean_norm_in_place(b)
print(b.to_list())
print()

print('znorm')
print(znorm(a).to_list())
print()

print('znorm_in_place')
b = a.copy()
znorm_in_place(b)
print(b.to_list())
print()

File: lib/khiva/normalization.rb
module Khiva
  module Normalization
    class << self
      def decimal_scaling_norm(tss)
        result = Utils.create_ptr
        FFI.call(:decimal_scaling_norm, tss, result)
        Array.new(result)
      end

      def decimal_scaling_norm!(tss)
        FFI.call(:decimal_scaling_norm_in_place, tss)
        tss
      end

      def max_min_norm(tss, high: 1, low: 0, epsilon: 0.00000001)
        result = Utils.create_ptr
        FFI.call(:max_min_norm, tss, Utils.double_ptr(high), Utils.double_ptr(low), Utils.double_ptr(epsilon), result)
        Array.new(result)
      end

      def max_min_norm!(tss, high: 1, low: 0, epsilon: 0.00000001)
        FFI.call(:max_min_norm_in_place, tss, Utils.double_ptr(high), Utils.double_ptr(low), Utils.double_ptr(epsilon))
        tss
      end

      def mean_norm(tss)
        result = Utils.create_ptr
        FFI.call(:mean_norm, tss, result)
        Array.new(result)
      end

      def mean_norm!(tss)
        FFI.call(:mean_norm_in_place, tss)
        tss
      end

      def znorm(tss, epsilon: 0.00000001)
        result = Utils.create_ptr
        FFI.call(:znorm, tss, Utils.double_ptr(epsilon), result)
        Array.new(result)
      end

      def znorm!(tss, epsilon: 0.00000001)
        FFI.call(:znorm_in_place, tss, Utils.double_ptr(epsilon))
        tss
      end
    end
  end
end
